enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,335 @@
+
+Akri, while an interesting approach that abstracts the Device Plugin framework,
+was deemed too complex for our requirements. The MicroShift team concluded that
+integrating Akri would significantly increase maintenance effort across
+engineering, QE, and documentation.
+
+Since the Generic Device Plugin (GDP) can be disabled via configuration, users
+retain the flexibility to utilize upstream Akri if their needs require it.
+
+## Open Questions [optional]
+
+## Test Plan
+
+Usually device plugins are tested using real hardware to make sure everything works.
+
+We'll need to get creative and implement some kind of dummy serial device to
+test communication between Pod utilizing the device and fake device
+(in the form of a daemon on the system).
+
+MicroShift's robust testing framework can host the integration tests, but it
+would be good to implement a simple test (like successful compilation with the 
+changes and start of MicroShift) in the GDP's fork repository.
+
+## Graduation Criteria
+
+Generic Device Plugin for MicroShift could start as GA.
+
+However, if we conclude that we're uncertain about "code import" method, we
+could publish GDP for MicroShift as Dev or Tech preview, gather the feedback,
+and decide if we want to transition to the "container" approach.
+
+### Dev Preview -> Tech Preview
+
+N/A
+
+### Tech Preview -> GA
+
+- End-to-end tests
+- End user documentation
+
+### Removing a deprecated feature
+
+N/A
+
+## Upgrade / Downgrade Strategy
+
+Generic Device Plugin will be part of MicroShift binary so there's no need for
+special consideration from user or technical point of view.
+
+When downgrading MicroShift to a version without Generic Device Plugin,
+the configuration fields of the new feature will be ignored by older binary.
+
+## Version Skew Strategy
+
+Generic Device Plugin fork will follow OpenShift's versioning and release schedule.
+
+However, it might not be required for MicroShift to track the fork GDP's HEAD of a branch
+to be up to date depending on the development velocity.
+For example, if by 4.25 we discover we need to make a fix in GDP, unless there
+were breaking changes between 4.20 and 4.25, we might only need to update the
+GDP's main branch and `go.mod` in each of the MicroShift's affected branches.
+
+## Operational Aspects of API Extensions
+
+Disabled GDP will not take any resources - the controller will not run.
+
+There are no resource consumption data for enabled embedded GDP
+(might depend on the amount of device presets in the configuration).
+The upstream manifest for GDP sets the following limits: 50m of CPU and 20Mi of memory.
+
+## Support Procedures
+
+If user created workload that is not starting, first step would be to `describe`
+the Pod or parent object to get the reason - what's preventing from scheduling
+or starting.
+```sh
+$ oc describe pod -n NAMESPACE POD
+$ oc describe deployment -n NAMESPACE DEPLOY
+$ oc describe statefulset -n NAMESPACE STATEFULSET
```

Comment:
do we need to enhance microshift sos report to include this stuff, also maybe the logs from the controllers? Or is this automatically all collected already?

Reply:
I believe the policy we have is that we don't collect workload information from non-microshift (i.e. user's) namespaces.
But the logs from controllers that are part of the microshift binary are in the journal

Reply:
fine!